File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/ApprovePinManifestInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Typed builder for the {@code ApprovePinManifest} instruction (SoraFS manifest lifecycle).
 * The recorded approval epoch comes exclusively from the block consensus timestamp.
 */
public final class ApprovePinManifestInstruction implements InstructionTemplate {

  public static final String ACTION = "ApprovePinManifest";

  private final String digestHex;
  private final String councilEnvelopeBase64;
  private final String councilEnvelopeDigestHex;
  private final Map<String, String> arguments;

  private ApprovePinManifestInstruction(final Builder builder) {
    this.digestHex = builder.digestHex;
    this.councilEnvelopeBase64 = builder.councilEnvelopeBase64;
    this.councilEnvelopeDigestHex = builder.councilEnvelopeDigestHex;
    this.arguments = Map.copyOf(builder.canonicalArguments());
  }

  public String digestHex() {
    return digestHex;
  }

  public String councilEnvelopeBase64() {
    return councilEnvelopeBase64;
  }

  public String councilEnvelopeDigestHex() {
    return councilEnvelopeDigestHex;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.CUSTOM;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static Builder builder() {
    return new Builder();
  }

  public static ApprovePinManifestInstruction fromArguments(final Map<String, String> arguments) {
    Objects.requireNonNull(arguments, "arguments");
    if (!ACTION.equals(arguments.get("action"))) {
      throw new IllegalArgumentException("Instruction argument 'action' must be " + ACTION);
    }
    for (final String key : arguments.keySet()) {
      if (!key.equals("action")
          && !key.equals("digest_hex")
          && !key.equals("council_envelope_base64")
          && !key.equals("council_envelope_digest_hex")) {
        throw new IllegalArgumentException("Unsupported ApprovePinManifest argument: " + key);
      }
    }
    final Builder builder = builder().setDigestHex(require(arguments, "digest_hex"));
    final String envelope = arguments.get("council_envelope_base64");
    if (envelope != null && !envelope.isBlank()) {
      builder.setCouncilEnvelopeBase64(envelope);
    }
    final String envelopeDigest = arguments.get("council_envelope_digest_hex");
    if (envelopeDigest != null && !envelopeDigest.isBlank()) {
      builder.setCouncilEnvelopeDigestHex(envelopeDigest);
    }
    return builder.build();
  }

  private static String require(final Map<String, String> arguments, final String key) {
    final String value = arguments.get(key);
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException("Instruction argument '" + key + "' is required");
    }
    return value;
  }

  private static String requireBase64(final String value, final String fieldName) {
    final String trimmed = Objects.requireNonNull(value, fieldName).trim();
    if (trimmed.isEmpty()) {
      throw new IllegalArgumentException(fieldName + " must not be blank");
    }
    final byte[] decoded;
    try {
      decoded = Base64.getDecoder().decode(trimmed);
    } catch (final IllegalArgumentException ex) {
      throw new IllegalArgumentException(fieldName + " must be base64", ex);
    }
    if (decoded.length == 0) {
      throw new IllegalArgumentException(fieldName + " must decode to non-empty bytes");
    }
    return trimmed;
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof ApprovePinManifestInstruction other)) {
      return false;
    }
    return Objects.equals(digestHex, other.digestHex)
        && Objects.equals(councilEnvelopeBase64, other.councilEnvelopeBase64)
        && Objects.equals(councilEnvelopeDigestHex, other.councilEnvelopeDigestHex);
  }

  @Override
  public int hashCode() {
    return Objects.hash(digestHex, councilEnvelopeBase64, councilEnvelopeDigestHex);
  }

  public static final class Builder {
    private String digestHex;
    private String councilEnvelopeBase64;
    private String councilEnvelopeDigestHex;

    private Builder() {}

    public Builder setDigestHex(final String digestHex) {
      this.digestHex = Objects.requireNonNull(digestHex, "digestHex");
      return this;
    }

    public Builder setCouncilEnvelopeBase64(final String councilEnvelopeBase64) {
      this.councilEnvelopeBase64 =
          requireBase64(councilEnvelopeBase64, "councilEnvelopeBase64");
      return this;
    }

    public Builder setCouncilEnvelopeBytes(final byte[] councilEnvelopeBytes) {
      Objects.requireNonNull(councilEnvelopeBytes, "councilEnvelopeBytes");
      return setCouncilEnvelopeBase64(
          Base64.getEncoder().encodeToString(councilEnvelopeBytes));
    }

    public Builder setCouncilEnvelopeDigestHex(final String digestHex) {
      this.councilEnvelopeDigestHex = Objects.requireNonNull(digestHex, "digestHex");
      return this;
    }

    public ApprovePinManifestInstruction build() {
      if (digestHex == null || digestHex.isBlank()) {
        throw new IllegalStateException("digestHex must be set");
      }
      return new ApprovePinManifestInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      args.put("digest_hex", digestHex);
      if (councilEnvelopeBase64 != null && !councilEnvelopeBase64.isBlank()) {
        args.put("council_envelope_base64", councilEnvelopeBase64);
      }
      if (councilEnvelopeDigestHex != null && !councilEnvelopeDigestHex.isBlank()) {
        args.put("council_envelope_digest_hex", councilEnvelopeDigestHex);
      }
      return args;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/ApprovePinManifestInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

import java.util.Base64

private const val APPROVE_PIN_ACTION = "ApprovePinManifest"

/**
 * Typed representation of the `ApprovePinManifest` instruction (SoraFS manifest lifecycle).
 */
class ApprovePinManifestInstruction private constructor(
    val digestHex: String,
    val approvedEpoch: Long,
    val councilEnvelopeBase64: String?,
    val councilEnvelopeDigestHex: String?,
    private val _arguments: Map<String, String>,
) : InstructionTemplate {

    override val kind: InstructionKind get() = InstructionKind.CUSTOM

    override val arguments: Map<String, String> get() = _arguments

    constructor(
        digestHex: String,
        approvedEpoch: Long,
        councilEnvelopeBase64: String? = null,
        councilEnvelopeDigestHex: String? = null,
    ) : this(
        digestHex = digestHex.also {
            require(it.isNotBlank()) { "digestHex must not be blank" }
        },
        approvedEpoch = approvedEpoch.also {
            require(it >= 0) { "approvedEpoch must be non-negative" }
        },
        councilEnvelopeBase64 = councilEnvelopeBase64?.let { requireBase64(it, "councilEnvelopeBase64") },
        councilEnvelopeDigestHex = councilEnvelopeDigestHex,
        _arguments = buildCanonicalArguments(digestHex, approvedEpoch, councilEnvelopeBase64, councilEnvelopeDigestHex),
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ApprovePinManifestInstruction) return false
        return approvedEpoch == other.approvedEpoch
            && digestHex == other.digestHex
            && councilEnvelopeBase64 == other.councilEnvelopeBase64
            && councilEnvelopeDigestHex == other.councilEnvelopeDigestHex
    }

    override fun hashCode(): Int {
        var result = digestHex.hashCode()
        result = 31 * result + approvedEpoch.hashCode()
        result = 31 * result + (councilEnvelopeBase64?.hashCode() ?: 0)
        result = 31 * result + (councilEnvelopeDigestHex?.hashCode() ?: 0)
        return result
    }

    companion object {
        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): ApprovePinManifestInstruction {
            val digestHex = require(arguments, "digest_hex")
            val approvedEpoch = requireLong(arguments, "approved_epoch")
            val envelope = arguments["council_envelope_base64"]?.takeIf { it.isNotBlank() }
            val envelopeDigest = arguments["council_envelope_digest_hex"]?.takeIf { it.isNotBlank() }
            return ApprovePinManifestInstruction(
                digestHex = digestHex,
                approvedEpoch = approvedEpoch,
                councilEnvelopeBase64 = envelope,
                councilEnvelopeDigestHex = envelopeDigest,
                _arguments = LinkedHashMap(arguments),
            )
        }

        @JvmStatic
        fun fromEnvelopeBytes(
            digestHex: String,
            approvedEpoch: Long,
            councilEnvelopeBytes: ByteArray,
            councilEnvelopeDigestHex: String? = null,
        ): ApprovePinManifestInstruction = ApprovePinManifestInstruction(
            digestHex = digestHex,
            approvedEpoch = approvedEpoch,
            councilEnvelopeBase64 = Base64.getEncoder().encodeToString(councilEnvelopeBytes),
            councilEnvelopeDigestHex = councilEnvelopeDigestHex,
        )

        private fun buildCanonicalArguments(
            digestHex: String,
            approvedEpoch: Long,
            councilEnvelopeBase64: String?,
            councilEnvelopeDigestHex: String?,
        ): Map<String, String> {
            val args = linkedMapOf(
                "action" to APPROVE_PIN_ACTION,
                "digest_hex" to digestHex,
                "approved_epoch" to approvedEpoch.toString(),
            )
            if (!councilEnvelopeBase64.isNullOrBlank()) {
                args["council_envelope_base64"] = councilEnvelopeBase64
            }
            if (!councilEnvelopeDigestHex.isNullOrBlank()) {
                args["council_envelope_digest_hex"] = councilEnvelopeDigestHex
            }
            return args
        }

        private fun require(arguments: Map<String, String>, key: String): String {
            val value = arguments[key]
            require(!value.isNullOrBlank()) { "Instruction argument '$key' is required" }
            return value
        }

        private fun requireLong(arguments: Map<String, String>, key: String): Long {
            val value = require(arguments, key)
            try {
                return value.toLong()
            } catch (ex: NumberFormatException) {
                throw IllegalArgumentException(
                    "Instruction argument '$key' must be a number: $value", ex,
                )
            }
        }

        private fun requireBase64(value: String, fieldName: String): String {
            val trimmed = value.trim()
            require(trimmed.isNotEmpty()) { "$fieldName must not be blank" }
            val decoded: ByteArray
            try {
                decoded = Base64.getDecoder().decode(trimmed)
            } catch (ex: IllegalArgumentException) {
                throw IllegalArgumentException("$fieldName must be base64", ex)
            }
            require(decoded.isNotEmpty()) { "$fieldName must decode to non-empty bytes" }
            return trimmed
        }
    }
}
